README.md
# RMagick

[![Build Status](https://travis-ci.org/gemhome/rmagick.svg?branch=master)](https://travis-ci.org/gemhome/rmagick)
[![Reference Status](https://www.versioneye.com/ruby/rmagick/reference_badge.svg)](https://www.versioneye.com/ruby/rmagick/references)

This is the official fork of RMagick. The decision to create a fork [originated here](https://github.com/rmagick/rmagick/issues/18#issuecomment-50022523), where you can read more about why it happened.

## Installation

Add the following to your `Gemfile`:
```ruby
gem 'rmagick'
```

Then run this:
```sh
bundle install
```

If you don't use a `Gemfile`, install it manually like this:
```sh
gem install rmagick
```

Require RMagick in your project as follows:
```ruby
require 'rmagick'
```


ChangeLog
RMagick 2.15.0
    o Ability to remove alpha channel - @ollie
    o C local variables guarded against GC to avoid segfaults - @u338steven
    o trace_proc protected with a mutex to avoid segfaults - @u338steven

RMagick 2.14.0
    o RMagick.rb moved to deprecated directory - @mockdeep
    o Better ImageMagick feature detection - @bf4
    o Prevent compilation failures if prefix is an empty string - @voxik
    o SplineInterpolatePixel preprocessor check removed - @joshk
    o Make error format a string literal - @mtasaka
    o Automatically set ARCHFLAGS on OS X - @u338steven
    o Fixed 'rmagick/version' require failure - @u338steven
    o Fixed escaping of '%' in Draw#get_type_metrics - @mkluczny
    o Silence warnings on Ruby 2.2 when comparing Enums - @meanphil
    o Multiple test suite improvements - @mockdeep, @bf4
    o Ruby source code formatting with RuboCop - @vassilevsky

RMagick 2.13.4
    #129 from @theschoolmaster
      proof of concept for using pkg-config in place of Magick-config on debian based systems

    #127 from @u338steven
      Changed Image#resample to calling ResampleImage (related #29, #45)

    #124 from @u338steven
      Fixed #122: lib/RMagick.rb is overwritten by lib/rmagick.rb on case-insensitive systems

    #123 from @prijutme4ty
      New class SolidFill in order to fill image with monochromatic background

    #121 from @markotom
      Quotes for correct path of font file

    #120 from @chulkilee
      Allow MagickCore6 from Magick-config

    #112 from @u338steven/
      Fixed: build error with ImageMagick 6.8.9 (when deprecated functions are excluded)

    #102 from @u338steven
      Fixed: related x_resolution, y_resolution

    from @u338steven
      Lots of test fixes

    #95 from @ioquatix
      Fix pixel hash test

    #94 from @u338steven
      Fixed: build error on Windows Ruby x64 (with ImageMagick 6.8.0-10 or Ima...

    #91 from @vassilevsky
      Do not test machine and OS-specific integers
      Fix initial memory limit test

RMagick 2.13.3
    #90 from @bricef
      Fix installation error on systems with HRDI enabled RMagick

RMagick 2.13.2
    o Fixed issues preventing RMagick from working with version 6.8 or higher
    o Fixed issues preventing RMagick from working with ruby 1.9.3

RMagick 2.13.1
    o Fixed bug preventing RMagick from working with version 6.5.9 or higher

RMagick 2.13.0
    o Added Doxygen documentation, for automatic documentation
    o Fixed bug #27467, get RMagick to compile witH ImageMagick 6.5.7
    o Fixed bug #27607, switch Pixel#from_hsla and Pixel#to_hsla to use ranges
      0-255 instead of 0-100 for saturation and lightness (range used by
      ImageMagick 6.5.6-5 and higher). Also added ability to specify all
      arguments to these functions as percentages (bug report by Arthur Chan).


RMagick 2.12.2
    o Add feature tests for SinusoidFunction and PolynomialFunction enum
      values to allow compiling with ImageMagick 6.4.8-7 (bug report by Mark
      Richman)

RMagick 2.12.1
    o Fix bug #27239, allow 2.12.0 to compile with older releases of ImageMagick
      (bug report by Sam Lown)

RMagick 2.12.0
    o Added Image#function_channel (available in ImageMagick 6.4.8-8)
    o Added Image#auto_level_channel, Image#auto_gamma_channel (available in
      ImageMagick 6.5.5-1)
    o Added Draw#interline_spacing, #interline_spacing= (available in
      ImageMagick 6.5.5-8)

RMagick 2.11.1
    o Applied Alexey Borzenkov's mingw patches to extconf.rb.
    o Fixed a bug in Magick.trace_proc that could cause a segv at program exit
      with Ruby 1.9.1 (bug report by Larry Young)
    o Added new CompressionType enum values ZipSCompression, PixCompression,
      Pxr24Compression, B44Compression, B44ACompression (available in
      ImageMagick 6.5.5-4)

RMagick 2.11.0
    o Fix bug #26475, dissolve and watermark don't work with new versions of
      ImageMagick (reported by Jim Crate)
    o Add Image#composite_mathematics (available in ImageMagick 6.5.4-3)
    o Add new LinearDodgeCompositeOp, LinearBurnCompositeOp,
      PegtopCompositeOp, PinLightCompositeOp, VividLightCompositeOp enum values
      (available in ImageMagick 6.5.4-3)

RMagick 2.10.0
    o ImageMagick releases earlier than 6.3.5-10 and Ruby releases earlier
      than 1.8.5 no longer supported.
    o (Experimental) Support the use of Ruby managed memory for all memory
      allocations (available in ImageMagick 6.5.3-10)
    o Add Image#selective_blur_channel (available in ImageMagick 6.5.0-3)
    o Add new AlphaBackgroundChannel enum value (available in ImageMagick
      6.5.2-5)
    o Add new DistortCompositeOp enum value (available in ImageMagick 6.5.3-7)

RMagick 2.9.2
    o Add new HorizontalTileEdgeVirtualPixelMethod,
      VerticalTileEdgeVirtualPixelMethod, CheckerTileVirtualPixelMethod
      VirtualPixelMethod enum values (available in ImageMagick 6.5.0-1)
    o Added BilinearForwardDistortion, BilinearReverseDistortion enums
      (available in ImageMagick 6.5.1-2)
    o Add missing composite operators to Magick::Draw#composite method
    o Add warning about dropping support for ImageMagick < 6.3.5 and
      Ruby < 1.8.5
    o Fix bug #25892, stack buffer overflow in Magick::TypeMetric.to_s
      (reported by Roman Simecek)

RMagick 2.9.1
    o Fix a bug that prevents the use of transparent background colors when
      built with ImageMagick 6.4.9-0

RMagick 2.9.0
    o Fix #23209, improve RVG's letter spacing (patch from Jonah Fox)
    o Add Draw#kerning= attribute (available in ImageMagick 6.4.7-8)
    o Add Draw#interword_spacing= attribute (available in ImageMagick
      6.4.8-0)
    o Add Draw#kerning, Draw#interword_spacing primitive methods (available in
      ImageMagick 6.4.8-3)
    o Feature #23171, support ImageList, Draw, Pixel marshaling.
    o Support all the new EvaluateOperator constants

RMagick 2.8.0
    o Add the endian, scene, and transparent_color attributes to Image::Info

examples/demo.rb
#!/usr/bin/env ruby -w
#
# Simple demo program for RMagick
#
# Concept and algorithms lifted from Magick++ demo script written
# by Bob Friesenhahn.
#
require 'rmagick'
include Magick

#
#   RMagick version of Magick++/demo/demo.cpp
#

Font = 'Helvetica'

begin
  puts 'Read images...'

  model = ImageList.new('../doc/ex/images/model.miff')
  model.border_color = 'black'
  model.background_color = 'black'
  model.cur_image[:Label] = 'RMagick'

  smile = ImageList.new('../doc/ex/images/smile.miff')
  smile.border_color = 'black'
  smile.cur_image[:Label] = 'Smile'

  #
  #   Create image stack
  #
  puts 'Creating thumbnails'

  # Construct an initial list containing five copies of a null
  # image. This will give us room to fit the logo at the top.
  # Notice I specify the width and height of the images via the
  # optional "size" attribute in the parm block associated with
  # the read method. There are two more examples of this, below.
  example = ImageList.new
  5.times { example.read('NULL:black') { self.size = '70x70'} }

  puts '   add noise...'
  example << model.add_noise(LaplacianNoise)
  example.cur_image[:Label] = 'Add Noise'

  puts '   annotate...'
  example << model.cur_image.copy
  example.cur_image[:Label] = 'Annotate'
  draw = Draw.new
  draw.annotate(example, 0, 0, 0, 20, 'RMagick') do
    self.pointsize = 18
    self.font = Font
    self.stroke = 'gold'
    self.fill = 'gold'
    self.gravity = NorthGravity
  end

  puts '   blur...'
  example << model.blur_image(0.0, 1.5)
  example.cur_image[:Label] = 'Blur'

  puts '   border...'
  example << model.border(6, 6, 'gold')
  example.cur_image[:Label] = 'Border'

  puts '   channel...'
  example << model.channel(RedChannel)
  example.cur_image[:Label] = 'Channel'

  puts '   charcoal...'
  example << model.charcoal
  example.cur_image[:Label] = 'Charcoal'

  puts '   composite...'
  example << model.composite(smile, 35, 65, OverCompositeOp)
  example.cur_image[:Label] = 'Composite'

  puts '   contrast...'
  example << model.contrast(false)
  example.cur_image[:Label] = 'Contrast'

  puts '   convolve...'
  kernel = [ 1, 1, 1, 1, 4, 1, 1, 1, 1 ]
  example << model.convolve(3, kernel)
  example.cur_image[:Label] = 'Convolve'

  puts '   crop...'
  example << model.crop(25, 50, 80, 80)
  example.cur_image[:Label] = 'Crop'

  puts '   despeckle...'
  example << model.despeckle
  example.cur_image[:Label] = 'Despeckle'

  puts '   draw...'
  example << model.cur_image.copy
  example.cur_image[:Label] = 'Draw'
  gc = Draw.new
  gc.fill 'black'
  gc.fill_opacity 0
  gc.stroke 'gold'
  gc.stroke_width 2
  gc.circle 60,90, 60,120
  gc.draw(example)

  puts '   edge...'
  example << model.edge(0)
  example.cur_image[:Label] = 'Detect Edges'

  puts '   emboss...'
  example << model.emboss
  example.cur_image[:Label] = 'Emboss'

  puts '   equalize...'
  example << model.equalize
  example.cur_image[:Label] = 'Equalize'

  puts '   explode...'
  example << model.implode(-1)
  example.background_color = '#000000ff'
  example.cur_image[:Label] = 'Explode'

  puts '   flip...'
  example << model.flip
  example.cur_image[:Label] = 'Flip'

  puts '   flop...'
  example << model.flop
  example.cur_image[:Label] = 'Flop'

  puts '   frame...'
  example << model.frame
  example.cur_image[:Label] = 'Frame'

  puts '   gamma...'
  example << model.gamma_correct(1.6)
  example.cur_image[:Label] = 'Gamma'

  puts '   gaussian blur...'
  example << model.gaussian_blur(1, 1.5)
  example.cur_image[:Label] = 'Gaussian Blur'

  # To add an Image in one of ImageMagick's built-in formats,
  # call the read method. The filename specifies the format and
  # any parameters it needs. The gradient format can be created in
  # any size. Specify the desired size by assigning it, in the form
  # "WxH", to the optional "size" attribute in the block associated
  # with the read method. Here we create a gradient image that is
  # the same size as the model image.
  puts '   gradient...'
  example.read('gradient:#20a0ff-#ffff00') do
    self.size = Geometry.new(model.columns, model.rows)
  end
  example.cur_image[:Label] = 'Gradient'

  puts '   grayscale...'
  example << model.cur_image.quantize(256, GRAYColorspace)
  example.cur_image[:Label] = 'Grayscale'

  puts '   implode...'
  example << model.implode(0.5)
  example.cur_image[:Label] = 'Implode'

  puts '   median filter...'
  example << model.median_filter(0)
  example.cur_image[:Label] = 'Median Filter'

  puts '   modulate...'
  example << model.modulate(1.10, 1.10, 1.10)
  example.cur_image[:Label] = 'Modulate'

  puts '   monochrome...'
  example << model.cur_image.quantize(2, GRAYColorspace, false)
  example.cur_image[:Label] = 'Monochrome'

  puts '   negate...'
  example << model.negate
  example.cur_image[:Label] = 'Negate'

  puts '   normalize...'
  example << model.normalize
  example.cur_image[:Label] = 'Normalize'

  puts '   oil paint...'
  example << model.oil_paint(3.0)
  example.cur_image[:Label] = 'Oil Paint'

  # The plasma format is very similar to the gradient format, above.
  puts '   plasma...'
  example.read('plasma:fractal') do
    self.size = Geometry.new(model.columns, model.rows)
  end
  example.cur_image[:Label] = 'Plasma'

  puts '   quantize...'
  example << model.cur_image.quantize
  example.cur_image[:Label] = 'Quantize'

  puts '   raise...'
  example << model.raise
  example.cur_image[:Label] = 'Raise'

  puts '   reduce noise...'
  example << model.reduce_noise(3.0)
  example.cur_image[:Label] = 'Reduce Noise'

  puts '   resize...'
  example << model.resize(0.50)
  example.cur_image[:Label] = 'Resize'

  puts '   roll...'
  example << model.roll(20, 10)
  example.cur_image[:Label] = 'Roll'

  puts '   rotate...'
  example << model.rotate(45).transparent('black')
  example.cur_image[:Label] = 'Rotate'

  puts '   scale...'
  example << model.scale(0.60)
  example.cur_image[:Label] = 'Scale'

  puts '   segment...'
  example << model.segment
  example.cur_image[:Label] = 'Segment'

  puts '   shade...'
  example << model.shade(false, 30, 30)
  example.cur_image[:Label] = 'Shade'

  puts '   sharpen...'
  example << model.sharpen(0.0, 1.0)
  example.cur_image[:Label] = 'Sharpen'

  puts '   shave...'
  example << model.shave(10, 10)
  example.cur_image[:Label] = 'Shave'

  puts '   shear...'
  example << model.shear(45, 45).transparent('black')
  example.cur_image[:Label] = 'Shear'

  puts '   spread...'
  example << model.spread(3)
  example.cur_image[:Label] = 'Spread'

  puts '   solarize...'
  example << model.solarize(50.0)
  example.cur_image[:Label] = 'Solarize'

  puts '   swirl...'
  temp = model.copy
  temp.background_color = '#000000ff'
  example << temp.swirl(90)
  example.cur_image[:Label] = 'Swirl'

  puts '   unsharp mask...'
  example << model.unsharp_mask(0.0, 1.0, 1.0, 0.05)
  example.cur_image[:Label] = 'Unsharp Mask'

  puts '   wave...'
  temp = model.copy
  temp.cur_image[:Label] = 'Wave'
  temp.matte = true
  temp.background_color = '#000000ff'
  example << temp.wave(25, 150)

  #
  #   Create image montage - notice the optional
  #   montage parameters are supplied via a block
  #

  puts 'Montage images...'

  montage = example.montage do
    self.geometry = '130x194+10+5>'
    self.gravity = CenterGravity
    self.border_width = 1
    rows = (example.size + 4) / 5
    self.tile = Geometry.new(5,rows)
    self.compose = OverCompositeOp

    # Use the ImageMagick built-in "granite" format
    # as the background texture.

    #       self.texture = Image.read("granite:").first
    self.background_color = 'white'
    self.font = Font
    self.pointsize = 18
    self.fill = '#600'
    self.filename = 'RMagick Demo'
    #       self.shadow = true
    #       self.frame = "20x20+4+4"
  end

  # Add the ImageMagick logo to the top of the montage. The "logo:"
  # format is a fixed-size image, so I don't need to specify a size.
  puts 'Adding logo image...'
  logo = Image.read('logo:').first
  if /GraphicsMagick/.match Magick_version
    logo.resize!(200.0/logo.rows)
  else
    logo.crop!(98, 0, 461, 455).resize!(0.45)
  end

  # Create a new Image for the composited montage and logo
  montage_image = ImageList.new
  montage_image << montage.composite(logo, 245, 0, OverCompositeOp)

  # Write the result to a file
  montage_image.compression = RLECompression
  montage_image.matte = false
  puts 'Writing image ./rm_demo_out.miff'
  montage_image.write 'rm_demo_out.miff'

# Uncomment the following lines to display image to screen
# puts "Displaying image..."
# montage_image.display

rescue
  puts "Caught exception: #{$ERROR_INFO}"
end

exit
